rfd/0156-notifications.md

```diff
@@ -0,0 +1,404 @@
+   // DefaultWarningCode is the default code for a warning notification.
+   DefaultWarningCode = "N0000W"
+
+   // ManuallyCreatedInformationalCode is the code for informational notifications that were manually created using `tctl`.
+   ManuallyCreatedInformationalCode = "N0001I"
+   // ManuallyCreatedWarningCode is the code for warning notifications that were manually created using `tctl`.
+   ManuallyCreatedWarningCode = "N0001W"
+
+   // AccessRequestReceivedCode is the code for the notification of a user receiving an access request.
+   AccessRequestReceivedCode = "N1000A"
+   // AccessRequestApprovedCode is the code for the notification of a user's access request being approved.
+   AccessRequestApprovedCode = "N1001A"
+   // AccessRequestDeniedCode is the code for the notification of a user's access request being denied.
+   AccessRequestDeniedCode = "N1001I"
+)
+```
+
+For example, if a notification's code is `N1000A`, the UI will make the notification clickable and redirect to the page for reviewing the access request.
+
+#### Usage
+
+The notifications service will work primarily by adding, deleting, and fetching notifications from the database. There will also be a watcher which
+watches for any changes to the database in order to perform real-time actions when a new notification is added.
+
+These are some examples of methods of the service which can be called:
+
+```go
+// Creates a user-specific notification for a user. This will be stored in the database as a Notification under `/notifications/user/<username>/<notification.id>`.
+NotifyUser(username string, notification types.Notification) error
+
+// Creates a global notification that matches users based on the specified role conditions (permissions). This will be stored in the database as a GlobalNotification under `/notifications/global/<notification.id>`.
+NotifyUsersByPermissions(roleConditions []types.RoleConditions, notification types.Notification, matchAllConditions bool) error
+
+// Creates a global notification that matches users who have the specified roles. This will be stored in the database as a GlobalNotification under `/notifications/global/<notification.id>`.
+NotifyUsersByRole(roles []string, notification types.Notification, matchAllConditions bool) error
+
+// Notifies all users. This will be stored in the database as a GlobalNotification under `/notifications/global/<notification.id>`.
+NotifyAllUsers(notification types.Notification) error
+
+// Fetches all notifications for the user, including all matching global notifications.
+// This will filter out notifications not meant for the user, as well as notifications the user has manually dismissed.
+// During this filtering, any notifications marked for expiry which are identified to be older than 30 days will be deleted from the database.
+// Additionally, any notification IDs found in the user's UserNotificationStates for notifications that no longer exist will be removed.
```

Comment:
Notifications that no longer exist, since when we are doing the filtering mentioned here, we would be dealing with the entire set of all notifications and not just the ones the user can see.